Add support NXP S32 HSE CRYPTO driver for S32Z270

haduongquang opened the pull request:
This PR introduces NXP S32 HSE CRYTO driver for SoC NXP S32Z27 and enables its usage for board s32z270dc2.
Supports cryptographic operations, including hashing and symmetric ciphers, with capabilities for ECB, CBC, and CTR modes using RAM-based key catalogs with 128-bit key lengths.

tests\crypto\crypto_hash:
```
SUITE PASS - 100.00% [crypto_hash]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.001 seconds
 PASS - [crypto_hash.test_hash] duration = 0.001 seconds
```

samples\drivers\crypto:
```
I: Cipher Sample
I: ECB Mode
I: Output length (encryption): 0
I: ECB mode ENCRYPT - Match
I: Output length (decryption): 0
I: ECB mode DECRYPT - Match
I: CBC Mode
I: Output length (encryption): 80
I: CBC mode ENCRYPT - Match
I: Output length (decryption): 0
I: CBC mode DECRYPT - Match
I: CTR Mode
I: Output length (encryption): 0
I: CTR mode ENCRYPT - Match
I: Output length (decryption): 0
I: CTR mode DECRYPT - Match
```

haduongquang commented:
Fix clang, compliance and build fail.

haduongquang commented:
Fixed compliance, clang and build fail.

Dat-NguyenDuy commented on drivers/crypto/crypto_nxp_hse.c:
```diff
@@ -0,0 +1,664 @@
+	}
+	default: {
+		return -ENOTSUP;
+	}
```
pls fix indentation

haduongquang replied:
not the wrong indentation, it is the correct format.

haduongquang commented on drivers/crypto/Kconfig.nxp_hse:
```diff
@@ -0,0 +1,30 @@
+
+config CRYPTO_NXP_HSE_FORMAT_KEY_CATALOG
+	bool "NXP HSE crypto driver supports formatting all key catalogs during initialization."
+	default n
```
I removed

haduongquang commented on drivers/crypto/Kconfig.nxp_hse:
```diff
@@ -0,0 +1,30 @@
+menuconfig CRYPTO_NXP_HSE
```
I updated kconfig, file name and driver also.

haduongquang commented on dts/arm/nxp/nxp_s32z27x_r52.dtsi:
```diff
@@ -1116,5 +1116,53 @@
 			status = "disabled";
 		};
+
+		mu0_mua: mu@23258000 {
```
I updated

haduongquang commented on drivers/crypto/crypto_nxp_hse.c:
```diff
@@ -0,0 +1,664 @@
+
+struct crypto_nxp_hse_data {
+	struct crypto_nxp_hse_session sessions[CONFIG_CRYPTO_NXP_HSE_MAX_SESSION];
+	uint8_t mu_instance;
```
I'm not using the config now, so I think we can keep it in data is better.

manuargue replied:
this is read only, it must be moved to the config

haduongquang commented on drivers/crypto/crypto_nxp_hse.c:
```diff
@@ -0,0 +1,664 @@
+	if (mu_channel != HSE_IP_INVALID_MU_CHANNEL_U8) {
+		session = &data->sessions[mu_channel - 1];
+		session->in_use = true;
+		session->channel = mu_channel;
```
I updated

haduongquang commented on drivers/crypto/crypto_nxp_hse.c:
```diff
@@ -0,0 +1,664 @@
+
+	session = ctx->drv_sessn_state;
+	k_mutex_lock(&session->crypto_lock, K_FOREVER);
+	memset(&session->req_type, 0, sizeof(Hse_Ip_ReqType));
```
To ensure the new description is correct, and unaffected by the old.

haduongquang commented on boards/nxp/s32z2xxdc2/doc/index.rst:
```diff
@@ -163,6 +165,19 @@ ADC is provided through ADC SAR controller with 2 instances. Each ADC SAR instan
+with 128-bit key lengths.
+
+.. note::
+   Applications must format key catalogs before executing the first service.
```
I documented it.

manuargue commented:
I'd suggest you to look at how the HSE was integrated into Linux. I think it's a cleaner approach than the current one proposed on this pr because it clearly shows that the MU's are coupled with the HSE and is explicit the reserved RAM to communicate between the host and the HSE.

https://github.com/nxp-auto-linux/linux/blob/49d9f8bb25e0de43587acd9238949752f914a6d2/arch/arm64/boot/dts/freescale/s32cc.dtsi#L596
https://github.com/nxp-auto-linux/linux/blob/49d9f8bb25e0de43587acd9238949752f914a6d2/arch/arm64/boot/dts/freescale/s32cc.dtsi#L105-L109

haduongquang commented on drivers/crypto/Kconfig.nxp_s32_hse:
```diff
@@ -0,0 +1,32 @@
+	default 0
+	depends on CRYPTO_NXP_S32_HSE
+	help
+	  The AES 128-bits Key Group ID within RAM Key Catalog.
```
I updated

haduongquang commented on samples/drivers/crypto/prj.conf:
```diff
@@ -3,4 +3,4 @@ CONFIG_CRYPTO_LOG_LEVEL_DBG=y
 CONFIG_LOG_MODE_MINIMAL=y
 
-CONFIG_MAIN_STACK_SIZE=4096
+CONFIG_MAIN_STACK_SIZE=8192
```
I reverted

haduongquang commented on drivers/crypto/crypto_nxp_s32_hse.c:
I have run the clang-format at locally, but some points are different with upstream.

manuargue replied:
https://github.com/zephyrproject-rtos/zephyr/commit/1f7c9605cb622bc2168d9a6aaca4d9a6cdc82de4

if it's not consistent, please report the version you are using in an issue

haduongquang replied:
I have fixed the clang-format.

haduongquang commented on drivers/crypto/crypto_nxp_s32_hse.c:
```diff
@@ -0,0 +1,661 @@
+
+#define CRYPTO_NXP_S32_HSE_SESSION_CFG(indx, _)                                                    \
+	{                                                                                          \
+		.channel = indx + 1, .out_buff = &crypto_out_buff[indx][0],                        \
```
I updated

haduongquang commented on drivers/crypto/crypto_nxp_s32_hse.c:
```diff
@@ -0,0 +1,661 @@
+#define CRYPTO_NXP_S32_HSE_MU_GET_INSTANCE(n)                                                      \
+	LISTIFY(__DEBRACKET HSE_IP_NUM_OF_MU_INSTANCES, CRYPTO_NXP_S32_HSE_MU_INSTANCE_CHECK, (|), n)
+
+#define CRYPTO_NXP_S32_HSE_BLOCK_KEY_LEN_BYTES 16U
```
I have added a Kconfig to do it

haduongquang commented on drivers/crypto/crypto_nxp_s32_hse.c:
```diff
@@ -0,0 +1,661 @@
+		return -EIO;
+	}
+
+	/* Update crypto crypto description input */
```
I updated

haduongquang commented on drivers/crypto/crypto_nxp_s32_hse.c:
```diff
@@ -0,0 +1,661 @@
+	Hse_Ip_ReqType crypto_req_type;
+	hseFormatKeyCatalogsSrv_t *format_key_serv =
+		&(crypto_serv_desc.hseSrv.formatKeyCatalogsReq);
+	uint8_t channel = Hse_Ip_GetFreeChannel(data->mu_instance);
```
I updated

haduongquang commented on drivers/crypto/crypto_nxp_s32_hse.c:
```diff
@@ -0,0 +1,661 @@
+
+static inline void free_session(struct crypto_nxp_s32_hse_session *session)
+{
+	session->in_use = false;
```
I updated

haduongquang commented on drivers/crypto/crypto_nxp_s32_hse.c:
```diff
@@ -0,0 +1,661 @@
+static struct crypto_nxp_s32_hse_session *crypto_nxp_s32_hse_get_session(const struct device *dev)
+{
+	struct crypto_nxp_s32_hse_data *data = dev->data;
+	uint8_t mu_channel = Hse_Ip_GetFreeChannel(data->mu_instance);
```
I updated.

haduongquang commented on drivers/crypto/Kconfig.nxp_s32_hse:
```diff
@@ -0,0 +1,32 @@
+	depends on CRYPTO_NXP_S32_HSE
+	help
+	  Enable supports formatting all key catalogs during initialization.
+	  This feature is primarily used for testing purposes.
```
I updated the samples/crypto to show how to do this.

manuargue replied:
> I updated the samples/crypto to show how to do this.

Having NXP specific HAL calls in a generic sample is definetly not a good idea. If this is really necessary it should be at least in `samples/boards/nxp`

ceolin replied:
Yep, that is a generic sample and should not have Hal specific calls.

haduongquang replied:
Thanks, I removed the HAL specific calls from the generic sample.

haduongquang commented:
```
@haduongquang I'd suggest you to look at how the HSE was integrated into Linux in regards to devicetree. I think it's a cleaner approach than the current one proposed on this pr because it clearly shows that the MU's are coupled with the HSE, and there's also explicit description of the RAM reserved to communicate between the CPU host and the HSE.

https://github.com/nxp-auto-linux/linux/blob/49d9f8bb25e0de43587acd9238949752f914a6d2/arch/arm64/boot/dts/freescale/s32cc.dtsi#L596
https://github.com/nxp-auto-linux/linux/blob/49d9f8bb25e0de43587acd9238949752f914a6d2/arch/arm64/boot/dts/freescale/s32cc.dtsi#L105-L109
````
For the first point, I have updated the device tree. 
For the second point, regarding `RAM reserved`. Some other platforms require the descriptor to be put in a dedicated RAM region in order to communicate with HSE, this is not the case for S32Z so a `RAM reserved` is not needed.

haduongquang commented:
> @haduongquang I'd suggest you to look at how the HSE was integrated into Linux in regards to devicetree. I think it's a cleaner approach than the current one proposed on this pr because it clearly shows that the MU's are coupled with the HSE, and there's also explicit description of the RAM reserved to communicate between the CPU host and the HSE.
> 
> * https://github.com/nxp-auto-linux/linux/blob/49d9f8bb25e0de43587acd9238949752f914a6d2/arch/arm64/boot/dts/freescale/s32cc.dtsi#L596
> * https://github.com/nxp-auto-linux/linux/blob/49d9f8bb25e0de43587acd9238949752f914a6d2/arch/arm64/boot/dts/freescale/s32cc.dtsi#L105-L109

For the first point, I have updated the device tree.
For the second point, regarding RAM reserved. Some other platforms require the descriptor to be put in a dedicated RAM region in order to communicate with HSE, this is not the case for S32Z so a RAM reserved is not needed.

manuargue commented on drivers/crypto/crypto_nxp_s32_hse.c:
```diff
@@ -0,0 +1,618 @@
+	struct crypto_nxp_s32_hse_data *data = dev->data;
+	struct crypto_nxp_s32_hse_session *session;
+
+	if (Hse_Ip_Init(data->mu_instance, &data->mu_state) != HSE_IP_STATUS_SUCCESS) {
```
It's not enough to just call this initialiation but also the MUs that are not active by default must be activated through a request to HSE. As it is right now implemented, if the user tries to use the Zephyr Crypto API on a HSE-MU device that is not active, it will fail. In general, the device probe function should ensure that all needed resources are allocated and ready in order for the driver to operate after exiting this function.

haduongquang replied:
The driver assumes that the HSE Firmware has already been installed and configured (including the activation of MU and configuration of key catalogs, etc.). Therefore, simply initializing MU is enough.

manuargue commented:
> @[mmahadevan108](https://github.com/mmahadevan108) mmahadevan108 requested review from [manuargue](https://github.com/manuargue) and [Dat-NguyenDuy](https://github.com/Dat-NguyenDuy) [8 hours ago](https://github.com/zephyrproject-rtos/zephyr/pull/79351#event-16019650571)

There are blocking comments not addressed.. awaiting for author to address them

haduongquang commented on boards/nxp/s32z2xxdc2/doc/index.rst:
```diff
@@ -163,6 +165,19 @@ ADC is provided through ADC SAR controller with 2 instances. Each ADC SAR instan
+.. note::
+   Applications must format key catalogs before executing the first service.
+
+   By default, MU0 is used, and other MUs are deactivated. If applications want to use other MUs,
```
This is out of date.

haduongquang commented:
> > @[mmahadevan108](https://github.com/mmahadevan108) mmahadevan108 requested review from [manuargue](https://github.com/manuargue) and [Dat-NguyenDuy](https://github.com/Dat-NguyenDuy) [8 hours ago](https://github.com/zephyrproject-rtos/zephyr/pull/79351#event-16019650571)
> 
> There are blocking comments not addressed.. awaiting for author to address them

Hi, I have addressed the comment and fixed the conflict also.

haduongquang commented:
I updated the copyright and improved and beautified the code. Fixed the typo in the commit message as well.

haduongquang commented:
Rebased and fixed the CI fail.

manuargue commented on drivers/crypto/Kconfig.nxp_s32_hse:
```diff
@@ -0,0 +1,33 @@
+if CRYPTO_NXP_S32_HSE
+
+config CRYPTO_NXP_S32_HSE_OUTPUT_BUFFER_SIZE
+	int "The output buffer size for storing the output data of HSE crypto service, measured in bytes."
```
```suggestion
	int "HSE crypto service output buffer size in bytes"
```

haduongquang replied:
I updated

manuargue commented on drivers/crypto/Kconfig.nxp_s32_hse:
```diff
@@ -0,0 +1,33 @@
+	  The output buffer size for storing the output data of HSE crypto service, measured in bytes.
+
+config CRYPTO_NXP_S32_HSE_AES_KEY_SIZE
+	int "The size of the key used in cryptographic algorithms, measured in bits."
```
```suggestion
	int "Key size in bits"
```

haduongquang replied:
I updated

manuargue commented on drivers/crypto/crypto_nxp_s32_hse.c:
```diff
@@ -0,0 +1,660 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(crypto_nxp_s32_hse);
+
+#define CRYPTO_NXP_S32_HSE_SERVICE_TIMEOUT 1000000000
```
```suggestion
#define CRYPTO_NXP_S32_HSE_SERVICE_TIMEOUT_MS 1000000000
```
to be consistent with the macro in line below (if it's not ms, use the corresponding time unit suffix)

haduongquang replied:
its unit is tick, I updated.

haduongquang commented on drivers/crypto/crypto_nxp_s32_hse.c:
```diff
@@ -0,0 +1,660 @@
+	if (mu_channel != HSE_IP_INVALID_MU_CHANNEL_U8) {
+		session = &data->sessions[mu_channel - 1];
+		session->in_use = true;
+		return session;
```
I updated

haduongquang commented on tests/crypto/crypto_hash/src/main.c:
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2022 Intel Corporation
+ * Copyright 2025 NXP
```
I removed

manuargue commented:
@haduongquang pls fix the twister failures

manuargue commented:
All my comments on this PR were addressed, but we have a blocker on the hal_nxp side that we are resolving internally at NXP. Once that's clarified, I will remove my block on this PR.

haduongquang committed:
boards: enables support for s32z270 hse crypto

Enable test for s32z270 hash crypto.

Enable samples for cipher cryptoEndable samples for EBC, CBC,
CTR mode of cipher crypto.

Signed-off-by: Ha Duong Quang <ha.duongquang@nxp.com>

manuargue reviewed (dismissed):
approving but we still need to merge https://github.com/zephyrproject-rtos/zephyr/pull/99190 to unblock the CI failures

manuargue commented:
> approving but we still need to merge #99190 to unblock the CI failures

@haduongquang the blocking pr was merged, pls rebase to prepare this to merge. Thanks

haduongquang committed:
driver: crypto: add NXP S32 CRYPTO HSE driver

Add device tree node for MU instances that will be used by HSE and RTU
for s32z270.

Add support hash crypto for NXP S32 with Algo 2:
SHA224, SHA256, SHA384 and SHA512.

Add support cipher crypto with ECB, CBC and CTR mode by using ram key
catalog.

Add support 128/256 bits ram key length.

Signed-off-by: Ha Duong Quang <ha.duongquang@nxp.com>

haduongquang commented:
> @haduongquang still manifest conflicts, pls rebase, that should clean the builds

I rebased

haduongquang commented on drivers/crypto/crypto_nxp_s32_hse.c:
```diff
@@ -0,0 +1,660 @@
+	return 0;
+}
+
+static struct crypto_driver_api crypto_nxp_s32_hse_api = {
```
I updated
